Sources/AsyncAlgorithms/zip/Zip2StateMachine.swift

```diff
@@ -0,0 +1,519 @@
+//
+//  Zip2StateMachine.swift
+//  
+//
+//  Created by Thibault Wittemberg on 13/09/2022.
+//
+
+final class Zip2Runtime<Base1: AsyncSequence, Base2: AsyncSequence>
+where Base1: Sendable, Base1.Element: Sendable, Base2: Sendable, Base2.Element: Sendable {
+  private let stateMachine = Zip2StateMachine<Base1.Element, Base2.Element>()
```

Comment:
Zip didn't need managed critical state before; it uses partial iteration with task groups instead

Reply:
I don’t think partial iteration applies here anymore since we only spawn a single Task<Void, Never> previously we spawned multiple Tasks which is unnecessary

Reply:
The StateMachine is now a struct and is wrapped inside a ManagedCriticalState inside the Zip2Runtime.